processor/worker.go

```diff
@@ -0,0 +1,211 @@
+package processor
+
+import (
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/rruntime"
+	"github.com/rudderlabs/rudder-server/services/rsources"
+	"github.com/rudderlabs/rudder-server/utils/logger"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+)
+
+// newWorker creates a new worker
+func newWorker(ctx context.Context, partition string, h workerHandle) *worker {
+	w := &worker{
+		handle:    h,
+		logger:    h.logger().Child(partition),
+		partition: partition,
+	}
+	w.lifecycle.ctx, w.lifecycle.cancel = context.WithCancel(ctx)
+	w.channel.ping = make(chan struct{}, 1)
+	w.channel.preproccess = make(chan subJob, w.handle.config().pipelineBufferedItems)
+	w.channel.transform = make(chan *transformationMessage, w.handle.config().pipelineBufferedItems)
+	w.channel.store = make(chan *storeMessage, (w.handle.config().pipelineBufferedItems+1)*(w.handle.config().maxEventsToProcess/w.handle.config().subJobSize+1))
+	w.start()
+
+	return w
+}
+
+// worker picks jobs from jobsdb for the appropriate partition and performs all processing steps for them
+//  1. preproccess
+//  2. transform
+//  3. store
+type worker struct {
+	partition string
+	handle    workerHandle
+	logger    logger.Logger
+
+	lifecycle struct { // worker lifecycle related fields
+		ctx    context.Context    // worker context
+		cancel context.CancelFunc // worker context cancel function
+		wg     sync.WaitGroup     // worker wait group
+
+		idleMu    sync.RWMutex // idle mutex
+		idleSince time.Time    // idle since
+	}
+
+	channel struct { // worker channels
+		ping        chan struct{}               // ping channel triggers the worker to start working
+		preproccess chan subJob                 // preproccess channel is used to send jobs to preproccess asynchronously when pipelining is enabled
+		transform   chan *transformationMessage // transform channel is used to send jobs to transform asynchronously when pipelining is enabled
+		store       chan *storeMessage          // store channel is used to send jobs to store asynchronously when pipelining is enabled
+	}
+}
+
+// start starts the various worker goroutines
+func (w *worker) start() {
+	// ping loop
+	w.lifecycle.wg.Add(1)
+	rruntime.Go(func() {
+		var exponentialSleep misc.ExponentialNumber[time.Duration]
+		defer w.lifecycle.wg.Done()
+		defer close(w.channel.preproccess)
+		defer close(w.channel.ping)
+		defer w.logger.Debugf("ping loop stopped for worker: %s", w.partition)
+		for {
+			select {
+			case <-w.lifecycle.ctx.Done():
+				return
+			case <-w.channel.ping:
+			}
+
+			w.setIdleSince(time.Time{})
+			if w.pickJobs() {
```

Comment:
Yes but I found it risky to introduce (i.e. respect) context in this step, since all our processes, so far, have been designed using a panic-driven approach, meaning that they cannot tolerate interruptions well.